llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -462,21 +541,35 @@ void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
 /// metadata and if so collects aspects from this metadata
 void processFunction(Function &F, FunctionToAspectsMapTy &FunctionToUsedAspects,
                      FunctionToAspectsMapTy &FunctionToDeclaredAspects,
-                     TypeToAspectsMapTy &TypesWithAspects, CallGraphTy &CG) {
+                     TypeToAspectsMapTy &TypesWithAspects, CallGraphTy &CG,
+                     const AspectValueToNameMapTy &AspectValues,
+                     bool FP64ConvEmu) {
+  auto FP64AspectIt = AspectValues.find("fp64");
+  assert(FP64AspectIt != AspectValues.end() &&
+         "fp64 aspect was not found in the aspect values.");
+  auto FP64Aspect = FP64AspectIt->second;
   const AspectsSetTy RetTyAspects =
       getAspectsFromType(F.getReturnType(), TypesWithAspects);
-  FunctionToUsedAspects[&F].insert(RetTyAspects.begin(), RetTyAspects.end());
+  for (const auto &Aspect : RetTyAspects)
+    if (!hasDoubleType(F.getReturnType()) || !FP64ConvEmu ||
+        (Aspect != FP64Aspect))
```

Comment:
The order of checks should be swapped here. We should first check  for `FP64ConvEmu` flag, then if the aspect is `fp64`, then and only then for use of `double` data type. This way "default" use case without emulation goes through the least amount of extra unnecessary checks. Same comment applies to other places where we have similar `if`